# Cleaning the eBay Car Listing Dataset

## About the data

The original Ebay car sales listing data set was from a German eBay website. The data set used in this notebook is a sampled data set which was uploaded to Kaggle and can be downloaded [here](https://www.kaggle.com/orgesleka/used-cars-database/data). 

Objective: Clean and analyze the used car sales listings

## Import packages and data

In [2]:
import pandas as pd
import numpy as np

In [3]:
autos = pd.read_csv("autos.csv", encoding='Latin-1')

## Data Cleaning

There are 5000 entries, 20 columns, and some columns have null values in this dataframe. The column names are in camelcase, to the Python preferred snakecase.

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

Interestingly, this dataframe includes 3 different time stamps in the columns: dateCrawled, dateCreated, and lastSeen.

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [5]:
#print out all the column names in autos dataframe
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
# convert column names from camelcase to snakecase
# create 'd' dictionary in which the keys are the old column names and values are the new column names 
d = {'dateCrawled':'date_crawled', 'offerType':'offer_type', 'vehicleType':'vehicle_type', 'yearOfRegistration':'year_of_registration', 'powerPS':'power_ps', 'monthOfRegistration':'month_of_registration', 'fuelType':'fuel_type', 'notRepairedDamage':'not_repaired_damage', 'dateCreated':'date_created', 'nrOfPictures':'nr_of_pictures', 'postalCode':'postal_code', 'lastSeen':'last_seen'}
autos = autos.rename(columns = d, inplace = False)

In [ ]:
# If your data set is large, use the code above to convert column names in camelcase to snakecase in your data set:
# initialize empty lists
upperCase = []
new_columns = []
lower_columns = []

# add the lower case column names to 'lower_columns' list
for column_name in autos.columns:
    if column_name.islower() == True:
        lower_columns.append(column_name)

# append the edited column names to the 'new_columns' list
for column_name in autos.columns:
    for c in column_name:
        upper_case = c.isupper()
        #append each uppercase letter in a column name to 'upperCase' list
        if upper_case == True:
            upperCase.append(c)
            #detect the uppercase in the column name that matches 
            #the uppercase letter in 'upperCase' list 
            for u in upperCase:
                if c == u in column_name:
                    #add an underscore in front of that uppercase letter
                    column_names = column_name.replace(u, "_" + u)
                    #make the uppercase letter to a lowercase letter
                    column_names = column_names.replace(u, u.lower())
                    #if there is a second uppercase letter in the column name,
                    #add an underscore in front of that uppercase letter
                    column_names = column_names.replace(u, "_" + u)
                    #make the uppercase letter to a lowercase letter
                    column_name = column_names.replace(u, u.lower())
                    #if the entire column name is lowercase
                    if column_name.islower() == True:
                        #add the column name to the 'new_columns' list
                        new_columns.append(column_name)

# Together the 'new_columns' and 'lower_columns' list with rename the 'autos' dataframe columns
autos.columns = new_columns + lower_columns  


## Inspect dataframe with edited column names

In [7]:
autos.describe()

year_of_registration      power_ps  month_of_registration  \
count          50000.000000  50000.000000           50000.000000   
mean            2005.073280    116.355920               5.723360   
std              105.712813    209.216627               3.711984   
min             1000.000000      0.000000               0.000000   
25%             1999.000000     70.000000               3.000000   
50%             2003.000000    105.000000               6.000000   
75%             2008.000000    150.000000               9.000000   
max             9999.000000  17700.000000              12.000000   

       nr_of_pictures   postal_code  
count         50000.0  50000.000000  
mean              0.0  50813.627300  
std               0.0  25779.747957  
min               0.0   1067.000000  
25%               0.0  30451.000000  
50%               0.0  49577.000000  
75%               0.0  71540.000000  
max               0.0  99998.000000

In [8]:
#inspect the data type of each column in autos dataframe
autos.dtypes

date_crawled             object
name                     object
seller                   object
offer_type               object
price                    object
abtest                   object
vehicle_type             object
year_of_registration      int64
gearbox                  object
power_ps                  int64
model                    object
odometer                 object
month_of_registration     int64
fuel_type                object
brand                    object
not_repaired_damage      object
date_created             object
nr_of_pictures            int64
postal_code               int64
last_seen                object
dtype: object

The `price` column has one string that needs to be replace with a NaN value and then needs to be saved as a numeric datatype. The `odometer` column name needs to include the unit of measurement used during data collection before turning the column into a numeric datatype.

In [5]:
#rename 'odometer' column as 'odometer_km'
autos = autos.rename(columns={"odometer": "odometer_km"})

In [6]:
#remove 'km' and comma string from a value in 'odometer_km' column
autos['odometer_km'] = autos['odometer_km'].str.replace('km','')
autos['odometer_km'] = autos['odometer_km'].str.replace(',','')
#convert 'odometer_km' column to integer data type
autos['odometer_km'] = autos['odometer_km'].astype(int)

In [7]:
#remove special character '$' and comma from a value in 'price' column
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
#convert 'price' column to integer data type
autos['price'] = autos['price'].astype(int)

Let's check that the changes were successful.

In [12]:
autos.describe()

price  year_of_registration      power_ps    odometer_km  \
count  5.000000e+04          50000.000000  50000.000000   50000.000000   
mean   9.840044e+03           2005.073280    116.355920  125732.700000   
std    4.811044e+05            105.712813    209.216627   40042.211706   
min    0.000000e+00           1000.000000      0.000000    5000.000000   
25%    1.100000e+03           1999.000000     70.000000  125000.000000   
50%    2.950000e+03           2003.000000    105.000000  150000.000000   
75%    7.200000e+03           2008.000000    150.000000  150000.000000   
max    1.000000e+08           9999.000000  17700.000000  150000.000000   

       month_of_registration  nr_of_pictures   postal_code  
count           50000.000000         50000.0  50000.000000  
mean                5.723360             0.0  50813.627300  
std                 3.711984             0.0  25779.747957  
min                 0.000000             0.0   1067.000000  
25%                 3.000000             0.0  30451.000000  
50%                 6.000000             0.0  49577.000000  
75%                 9.000000             0.0  71540.000000  
max                12.000000             0.0  99998.000000

In [157]:
autos.dtypes

date_crawled             object
name                     object
seller                   object
offer_type               object
price                     int64
abtest                   object
vehicle_type             object
year_of_registration      int64
gearbox                  object
power_ps                  int64
model                    object
odometer_km               int64
month_of_registration     int64
fuel_type                object
brand                    object
not_repaired_damage      object
date_created             object
nr_of_pictures            int64
postal_code               int64
last_seen                object
dtype: object

## Removing outliers

Analyze the columns for values that are unrealistically high or low (outliers) and remove those values from the dataframe. Let's look at `price` and `odometer_km` columns.

Let's explore the price column.

In [8]:
#show unique values in the 'price' column
autos["price"].unique().shape

(2357,)

In [38]:
#view the min, max, median, mean, etc
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [42]:
#shows the highest value counts for 'price' column
autos["price"].value_counts().head()

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64

In [40]:
#shows lowest value counts for 'price' column
autos["price"].value_counts().tail()

20790    1
8970     1
846      1
2895     1
33980    1
Name: price, dtype: int64

In [34]:
#shows the lowest indexes with their values for 'price' column
autos["price"].sort_index(ascending=True).head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [35]:
#shows the highest indices with their values for 'price' column
autos["price"].sort_index(ascending=False).head()

49999     1250
49998    22900
49997    13200
49996     1980
49995    24900
Name: price, dtype: int64

Let's explore the odometer_km column.

In [43]:
#show unique values in the 'odometer_km' column
autos["odometer_km"].unique().shape

(13,)

In [15]:
#view the min, max, median, mean, etc
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [48]:
#shows the highest value counts for 'odometer_km' column
autos["odometer_km"].value_counts().head()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64

In [49]:
#shows the lowest value counts for 'odometer_km' column
autos["odometer_km"].value_counts().tail()

5000     967
40000    819
30000    789
20000    784
10000    264
Name: odometer_km, dtype: int64

In [50]:
#shows the lowest indices with their values for 'odometer_km' column
autos["odometer_km"].sort_index(ascending=True).head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [51]:
#shows the highest indices with their values for 'odometer_km' column
autos["odometer_km"].sort_index(ascending=False).head()

49999    150000
49998     40000
49997      5000
49996    150000
49995    100000
Name: odometer_km, dtype: int64

The price column has zeros that need to be removed before continuing the analysis.

In [9]:
#removing zeros from 'price' column in autos dataframe
autos = autos[autos['price'].between(1,350000)]

Now, let's look at 5 columns that have date values. It is important to note that the date information came from two sources: crawler and the website itself. The crawler provide date information for the `date_crawled` and `last_seen` columns. The website provided the date information for the `registration_month`, and `registration_year` columns.

The `registration_month`, and `registration_year` columns have date information in a numerical format. 

The `date_crawled` and `last_seen` columns have date information in string format.

Let's see how to convert this information to a numerical format for the `date_crawled` column.

In [10]:
#extract date values from 'date_crawled' column
date_crawled_date_values = autos['date_crawled'].str[:10]
#include missing values in the distribution
#use percentage instead of counts
date_crawled_dates = date_crawled_date_values.value_counts(normalize=True, dropna=False)
#rank by date in ascending order (earliest to latest)
date_crawled = date_crawled_dates.sort_index()
#print series
print(date_crawled)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64


Let's see how to convert this information to a numerical format for the `last_seen` column.

In [11]:
#extract date values from 'last_seen' column
last_seen_date_values = autos['last_seen'].str[:10]
#include missing values in the distribution
#use percentage instead of counts
last_seen_dates = last_seen_date_values.value_counts(normalize=True, dropna=False)
#rank by date in ascending order (earliest to latest)
last_seen = last_seen_dates.sort_index()
#print series
print(last_seen)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


## Handling the incorrect Registation Year data

Let's explore the distribution for the `registration_year` column.

The current minimum registration year of 1000 is inaccurate because it is before cars were invented. The current maximum registration year of 9999 is inaccurate because it is thousands of years in the future. A new maximum and minimum for the registration year column need to be established.

In [12]:
#describe registration year column
autos['year_of_registration'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64

After examining the registration years before 1900 in the autos dataframe, the lower limit of 1900 is reasonable since the maximum value for the registration year before 1900 is 1800 which is before cars were invented.

In [13]:
autos_before_1900 = autos[autos['year_of_registration'] < 1900]

In [14]:
autos_before_1900['year_of_registration'].describe()

count       5.000000
mean     1342.400000
std       420.158661
min      1000.000000
25%      1001.000000
50%      1111.000000
75%      1800.000000
max      1800.000000
Name: year_of_registration, dtype: float64

After examining the registration years after 2016 in the autos dataframe, the hhighest value for the registration year in the 2000s is 2019. All years greater than 2019 not reasonable since they are too far into the future.

In [49]:
autos_after_2016 = autos[autos['year_of_registration'] > 2016]

In [50]:
autos_after_2016['year_of_registration'].describe()

count    1864.000000
mean     2017.254292
std         0.438037
min      2017.000000
25%      2017.000000
50%      2017.000000
75%      2018.000000
max      2019.000000
Name: year_of_registration, dtype: float64

In [17]:
autos_after_2016['year_of_registration'].value_counts()

2017    1392
2018     470
5000       4
9999       3
2019       2
5911       1
9000       1
2800       1
4800       1
8888       1
4500       1
6200       1
4100       1
Name: year_of_registration, dtype: int64

Let's remove the inaccurate registration years and calcuate the distribution of the column!

In [18]:
#removing inaccurate years from 'year_of_registration' column in autos dataframe
autos = autos[autos['year_of_registration'].between(1900,2019)]

In [19]:
#calculate the distribution
autos['year_of_registration'].value_counts(normalize=True)

2000    0.065012
2005    0.060480
1999    0.059677
2004    0.055680
2003    0.055598
2006    0.055001
2001    0.054300
2002    0.051210
1998    0.048676
2007    0.046905
2008    0.045628
2009    0.042950
1997    0.040190
2011    0.033433
2010    0.032733
2017    0.028674
1996    0.028283
2012    0.026985
1995    0.025276
2016    0.025131
2013    0.016541
2014    0.013657
1994    0.012957
2018    0.009682
1993    0.008755
2015    0.008075
1992    0.007622
1990    0.007148
1991    0.006983
1989    0.003584
          ...   
1976    0.000433
1969    0.000391
1975    0.000371
1965    0.000350
1964    0.000247
1963    0.000165
1959    0.000124
1961    0.000124
1910    0.000103
1937    0.000082
1962    0.000082
1958    0.000082
1956    0.000082
1950    0.000062
1955    0.000041
2019    0.000041
1934    0.000041
1957    0.000041
1951    0.000041
1941    0.000041
1954    0.000041
1953    0.000021
1948    0.000021
1927    0.000021
1943    0.000021
1929    0.000021
1939    0.000021
1938    0.0000

## Calculate the mean prices for the top 10 auto brands in the autos dataframe

In [69]:
#view the unique brands in autos dataframe and their counts
auto_brands = autos['brand'].value_counts()
#select top 10 most popular auto brands and their counts
top_10 = auto_brands[0:10]
#select top 10 auto brands
top_10_brands = top_10.index
#create an empty dictionary
brand_mean_prices = {}
#for each brand in the top 10 brands list
for brand in top_10_brands:
    #assign the brand name as the key and the brand's mean price as the value
    brand_mean_prices[brand] = autos.loc[autos['brand'] == brand, 'price'].mean()
    
#List the top 10 most expensive auto brands and their mean prices from greatest to least
sorted_brands = sorted(brand_mean_prices, key=brand_mean_prices.get, reverse=True)
for brand in sorted_brands:
    print(brand, brand_mean_prices[brand])

audi 9212.9306621881
mercedes_benz 8526.623225806452
bmw 8261.382442169132
volkswagen 5333.1962055948115
seat 4320.168661588684
ford 3728.4121821407452
peugeot 3065.611888111888
opel 2941.4664391353813
fiat 2793.8700475435817
renault 2431.195698924731


## Calculate the mean mileage for the top 10 auto brands in the autos dataframe

In [72]:
#view the unique brands in autos dataframe and their counts
auto_brands = autos['brand'].value_counts()
#select top 10 most popular auto brands and their counts
top_10 = auto_brands[0:10]
#select top 10 auto brands
top_10_brands = top_10.index
#create an empty dictionary
brand_mean_mileage = {}
#for each brand in the top 10 brands list
for brand in top_10_brands:
    #assign the brand name as the key and the brand's mean mileage as the value
    brand_mean_mileage[brand] = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()
    
#List the top 10 most expensive auto brands and their mean mileage from greatest to least
sorted_mileage = sorted(brand_mean_mileage, key=brand_mean_mileage.get, reverse=True)
for brand in sorted_mileage:
    print(brand, brand_mean_mileage[brand])

bmw 132682.97307546454
mercedes_benz 130848.3870967742
audi 129492.56238003839
opel 129452.02882062951
volkswagen 128955.570612719
renault 128062.36559139784
peugeot 127356.64335664336
ford 124349.49733885274
seat 121768.22633297062
fiat 117567.35340729002


## Combine mean prices and mean mileage dictionaries into a dataframe

In [76]:
#convert mean prices dictionary to Series object
bmp_series = pd.Series(brand_mean_prices)
#convert mean milages dictionary to Series object
bmm_series = pd.Series(brand_mean_mileage)
#create 'df' dataframe from mean prices Series object
df = pd.DataFrame(bmp_series, columns=['mean_prices'])
#add mean mileage Series object to new dataframe 'df'
df['mean_mileage'] = bmm_series

In [77]:
#pretty print 'df' dataframe
df

mean_prices   mean_mileage
audi           9212.930662  129492.562380
bmw            8261.382442  132682.973075
fiat           2793.870048  117567.353407
ford           3728.412182  124349.497339
mercedes_benz  8526.623226  130848.387097
opel           2941.466439  129452.028821
peugeot        3065.611888  127356.643357
renault        2431.195699  128062.365591
seat           4320.168662  121768.226333
volkswagen     5333.196206  128955.570613